src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,225 @@
+-- | These functions allow PureScript strings to be treated as if they were
+-- | sequences of Unicode code points instead of their true underlying
+-- | implementation (sequences of UTF-16 code units). For nearly all uses of
+-- | strings, these functions should be preferred over the ones in Data.String.
+module Data.String.CodePoints
+  ( module StringReExports
+  , CodePoint()
+  , codePointAt
+  , codePointFromInt
+  , codePointToInt
+  , count
+  , drop
+  , dropWhile
+  , fromCodePointArray
+  , indexOf
+  , indexOf'
+  , lastIndexOf
+  , lastIndexOf'
+  , length
+  , singleton
+  , splitAt
+  , take
+  , takeWhile
+  , toCodePointArray
+  , uncons
+  ) where
+
+import Prelude
+
+import Data.Array as Array
+import Data.Char as Char
+import Data.List (List(Cons, Nil), fromFoldable)
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Newtype (class Newtype)
+import Data.String as String
+import Data.String.Unsafe as Unsafe
+-- WARN: This list must be updated to re-export any exports added to Data.String. That makes me sad.
+import Data.String (Pattern(..), Replacement(..), charAt, charCodeAt, contains, fromCharArray, joinWith, localeCompare, null, replace, replaceAll, split, stripPrefix, stripSuffix, toChar, toCharArray, toLower, toUpper, trim) as StringReExports
+import Data.Tuple (Tuple(Tuple))
+import Data.Unfoldable (unfoldr)
+
+
+-- | CodePoint is an Int bounded between 0 and 0x10FFFF, corresponding to
+-- | Unicode code points.
+newtype CodePoint = CodePoint Int
+
+derive instance eqCodePoint :: Eq CodePoint
+derive instance ordCodePoint :: Ord CodePoint
+derive instance newtypeCodePoint :: Newtype CodePoint _
```

Comment:
I just realised that defining this instance means that people can create `CodePoint` values larger than `0x10FFFF` or smaller than 0. Presumably that's a problem?